File: scripts/skillopt-eval/contracts/workflow.ts
import { z } from "zod";
import {
  ArtifactIdSchema,
  CONTRACT_SCHEMA_VERSION,
  NonEmptyStringSchema,
  PriceEquivalentEstimateSchema,
  Sha256Schema,
  TimestampSchema,
  UsageSchema,
  boundedContractSchema,
} from "./common";
import { SkillSchema, VariantSchema } from "./episode";

// implements REQ-skillopt-codex-optimization
export const LedgerEntrySchema = boundedContractSchema(
  z
    .object({
      schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
      artifactType: z.literal("ledger-entry"),
      runId: ArtifactIdSchema,
      sequence: z.int().nonnegative(),
      previousEntryHash: Sha256Schema.nullable(),
      entryHash: Sha256Schema,
      occurredAt: TimestampSchema,
      episodeId: ArtifactIdSchema.optional(),
      category: z.enum([
        "preflight",
        "development",
        "optimization",
        "held-out",
        "bundle",
        "infrastructure",
      ]),
      model: z.enum(["gpt-5.4-mini", "gpt-5.5", "none"]),
      usage: UsageSchema,
      priceEquivalentEstimate: PriceEquivalentEstimateSchema,
    })
    .strict()
    .superRefine((entry, context) => {
      const validLink =
        entry.sequence === 0
          ? entry.previousEntryHash === null
          : entry.previousEntryHash !== null;
      if (!validLink)
        context.addIssue({
          code: "custom",
          message: "ledger sequence/link mismatch",
        });
    }),
);

// implements REQ-skillopt-codex-optimization
export const RunStateSchema = boundedContractSchema(
  z
    .object({
      schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
      artifactType: z.literal("run-state"),
      runId: ArtifactIdSchema,
      runLockHash: Sha256Schema,
      phase: z.enum([
        "preflight",
        "development",
        "optimization",
        "held-out",
        "bundle",
        "review",
        "complete",
        "no-go",
      ]),
      completedEpisodeIds: z
        .array(ArtifactIdSchema)
        .refine(
          (ids) => new Set(ids).size === ids.length,
          "completed episode ids must be unique",
        ),
      ledgerHeadHash: Sha256Schema.nullable(),
      updatedAt: TimestampSchema,
      interrupted: z.boolean(),
    })
    .strict()
    .superRefine((state, context) => {
      if (state.phase === "complete" && state.interrupted) {
        context.addIssue({
          code: "custom",
          message: "complete run state cannot be interrupted",
        });
      }
    }),
);

const LegacyReportSchema = boundedContractSchema(
  z
    .object({
      runId: NonEmptyStringSchema,
      skill: NonEmptyStringSchema,
      variants: z.tuple([
        z.literal("baseline"),
        z.literal("one-shot"),
        z.literal("skillopt"),
      ]),
      cells: z.array(z.json()).min(1),
      costUsd: z.number().min(0).max(400),
      verdict: z.enum(["pass", "fail", "no-go"]),
    })
    .strict(),
);

// implements REQ-skillopt-codex-optimization
export const ReportV1Schema = boundedContractSchema(
  z
    .object({
      schemaVersion: z.literal(CONTRACT_SCHEMA_VERSION),
      artifactType: z.literal("report"),
      runId: NonEmptyStringSchema,
      runLockHash: Sha256Schema,
      skill: SkillSchema,
      variants: z
        .tuple([VariantSchema, VariantSchema, VariantSchema])
        .refine(
          (variants) =>
            variants[0] === "baseline" &&
            variants[1] === "one-shot" &&
            variants[2] === "skillopt",
          "variants must use canonical order",
        ),
      cells: z
        .array(Sha256Schema)
        .min(1)
        .refine(
          (cells) => new Set(cells).size === cells.length,
          "cells must be unique",
        ),
      priceEquivalentEstimate: PriceEquivalentEstimateSchema,
      verdict: z.enum(["pass", "fail", "no-go"]),
      generatedAt: TimestampSchema,
      gateResults: z
        .object({
          aggregate: z.boolean(),
          bootstrap: z.boolean(),
          family: z.boolean(),
          security: z.boolean(),
          bundle: z.boolean().nullable(),
        })
        .strict(),
    })
    .strict(),
);

// implements REQ-skillopt-codex-optimization
export const ReportSchema = boundedContractSchema(
  z.union([ReportV1Schema, LegacyReportSchema]),
);
// implements REQ-skillopt-codex-optimization
export type Report = Readonly<z.infer<typeof ReportSchema>>;

File: tools/skillopt/kibi_skillopt/workflow.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import Field, field_validator, model_validator
from typing_extensions import Self

from .common import (
    ArtifactId,
    ContractModel,
    ContractValidationError,
    JsonBoolean,
    JsonInteger,
    JsonNode,
    JsonNumber,
    JsonValue,
    NonEmptyString,
    PriceEquivalentEstimate,
    Sha256,
    Timestamp,
    Usage,
)
from .episode import Skill


class LedgerEntry(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["ledger-entry"], Field(alias="artifactType")]
    run_id: Annotated[ArtifactId, Field(alias="runId")]
    sequence: Annotated[JsonInteger, Field(ge=0)]
    previous_entry_hash: Annotated[Sha256 | None, Field(alias="previousEntryHash")]
    entry_hash: Annotated[Sha256, Field(alias="entryHash")]
    occurred_at: Annotated[Timestamp, Field(alias="occurredAt")]
    episode_id: Annotated[ArtifactId | None, Field(alias="episodeId")] = None
    category: Literal[
        "preflight", "development", "optimization", "held-out", "bundle", "infrastructure"
    ]
    model: Literal["gpt-5.4-mini", "gpt-5.5", "none"]
    usage: Usage
    price_equivalent_estimate: Annotated[
        PriceEquivalentEstimate, Field(alias="priceEquivalentEstimate")
    ]

    @field_validator("episode_id", mode="before")
    @classmethod
    def reject_explicit_null_episode_id(cls, value: JsonValue) -> JsonValue:
        if value is None:
            raise ContractValidationError("episodeId cannot be null")
        return value

    @model_validator(mode="after")
    def verify_chain_link(self) -> Self:
        valid_link = (
            self.previous_entry_hash is None
            if self.sequence == 0
            else self.previous_entry_hash is not None
        )
        if not valid_link:
            raise ContractValidationError("ledger sequence/link mismatch")
        return self


class RunState(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["run-state"], Field(alias="artifactType")]
    run_id: Annotated[ArtifactId, Field(alias="runId")]
    run_lock_hash: Annotated[Sha256, Field(alias="runLockHash")]
    phase: Literal[
        "preflight",
        "development",
        "optimization",
        "held-out",
        "bundle",
        "review",
        "complete",
        "no-go",
    ]
    completed_episode_ids: Annotated[tuple[ArtifactId, ...], Field(alias="completedEpisodeIds")]
    ledger_head_hash: Annotated[Sha256 | None, Field(alias="ledgerHeadHash")]
    updated_at: Annotated[Timestamp, Field(alias="updatedAt")]
    interrupted: JsonBoolean

    @model_validator(mode="after")
    def verify_unique_completed_episodes(self) -> Self:
        if len(set(self.completed_episode_ids)) != len(self.completed_episode_ids):
            raise ContractValidationError("completed episode ids must be unique")
        if self.phase == "complete" and self.interrupted:
            raise ContractValidationError("complete run state cannot be interrupted")
        return self


class LegacyReport(ContractModel):
    run_id: Annotated[NonEmptyString, Field(alias="runId")]
    skill: NonEmptyString
    variants: tuple[Literal["baseline"], Literal["one-shot"], Literal["skillopt"]]
    cells: tuple[JsonNode, ...]
    cost_usd: Annotated[JsonNumber, Field(alias="costUsd", ge=0, le=400)]
    verdict: Literal["pass", "fail", "no-go"]


class GateResults(ContractModel):
    aggregate: JsonBoolean
    bootstrap: JsonBoolean
    family: JsonBoolean
    security: JsonBoolean
    bundle: JsonBoolean | None


class ReportV1(ContractModel):
    schema_version: Annotated[Literal["1.0.0"], Field(alias="schemaVersion")]
    artifact_type: Annotated[Literal["report"], Field(alias="artifactType")]
    run_id: Annotated[ArtifactId, Field(alias="runId")]
    run_lock_hash: Annotated[Sha256, Field(alias="runLockHash")]
    skill: Skill
    variants: tuple[Literal["baseline"], Literal["one-shot"], Literal["skillopt"]]
    cells: Annotated[tuple[Sha256, ...], Field(min_length=1)]
    price_equivalent_estimate: Annotated[
        PriceEquivalentEstimate, Field(alias="priceEquivalentEstimate")
    ]
    verdict: Literal["pass", "fail", "no-go"]
    generated_at: Annotated[Timestamp, Field(alias="generatedAt")]
    gate_results: Annotated[GateResults, Field(alias="gateResults")]

    @model_validator(mode="after")
    def verify_unique_cells(self) -> Self:
        if len(set(self.cells)) != len(self.cells):
            raise ContractValidationError("cells must be unique")
        return self
